src/main/java/com/cloudbees/jenkins/plugins/bitbucket/hooks/PushHookProcessor.java

```diff
@@ -44,10 +68,124 @@ public void process(String payload, BitbucketType instanceType) {
             }
             if (push != null) {
                 String owner = push.getRepository().getOwnerName();
-                String repository = push.getRepository().getRepositoryName();
+                final String repository = push.getRepository().getRepositoryName();
+                if (push.getChanges().isEmpty()) {
+                    LOGGER.log(Level.INFO, "Received hook from Bitbucket. Processing push event on {0}/{1}",
+                            new Object[]{owner, repository});
+                    scmSourceReIndex(owner, repository);
+                } else {
+                    SCMHeadEvent.Type type = null;
+                    for (BitbucketPushEvent.Change change: push.getChanges()) {
+                        if ((type == null || type == SCMEvent.Type.CREATED) && change.isCreated()) {
+                            type = SCMEvent.Type.CREATED;
+                        } else if ((type == null || type == SCMEvent.Type.REMOVED) && change.isClosed()) {
+                            type = SCMEvent.Type.REMOVED;
+                        } else {
+                            type = SCMEvent.Type.UPDATED;
+                        }
+                    }
+                    SCMHeadEvent.fireNow(new SCMHeadEvent<BitbucketPushEvent>(type, push, origin) {
+                        @Override
+                        public boolean isMatch(@NonNull SCMNavigator navigator) {
```

Comment:
Perhaps, but for now I am ok with the duplication as fixing it would require establishing commonality between `BitbucketPushEvent` and `BitbucketPullRequestEvent`